src/Microsoft.Extensions.Options/Options.cs

```diff
@@ -1,6 +1,71 @@
-﻿namespace Microsoft.Extensions.Options
+﻿using System;
+
+namespace Microsoft.Extensions.Options
 {
     /// <summary>
+    /// Implementation of IOptionsFactory.
+    /// </summary>
+    /// <typeparam name="TOptions">The type of options being requested.</typeparam>
+    public class OptionsService<TOptions> : IOptions<TOptions>, IOptionsSnapshot<TOptions> where TOptions : class, new()
+    {
+        private readonly IOptionsCache<TOptions> _cache;
+        private readonly IOptionsFactory<TOptions> _factory;
+
+        /// <summary>
+        /// Initializes a new instance with the specified options configurations.
+        /// </summary>
+        /// <param name="cache">The cache to use.</param>
+        /// <param name="factory">The factory to use to create options.</param>
+        public OptionsService(IOptionsCache<TOptions> cache, IOptionsFactory<TOptions> factory)
+        {
+            _cache = cache;
+            _factory = factory;
+        }
+
+        public TOptions Value
+        {
+            get
+            {
+                return Get("");
+            }
+        }
+
+        public virtual void Add(string name, TOptions options)
+        {
+            if (name == null)
+            {
+                throw new ArgumentNullException(nameof(name));
+            }
+            if (options == null)
+            {
+                throw new ArgumentNullException(nameof(options));
+            }
+            if (!_cache.TryAdd(name, options))
+            {
+                throw new InvalidOperationException("An option named {name} already exists.");
+            }
+        }
+
+        public virtual TOptions Get(string name)
+        {
+            if (name == null)
+            {
+                throw new ArgumentNullException(nameof(name));
+            }
+            return _cache.GetOrAdd(name, () => _factory.Create(name));
```

Comment:
`IOptionsSource` that you mentioned looks the same like `IOptions.Value` implementation with injected factory. Factory starts working only when we access `Value`.

The other case if we need `Task<TOptions> ValueAsync` member. Extend the current interface? or create the inherited (`IOptionsAsync`) one (for advanced users/cases).